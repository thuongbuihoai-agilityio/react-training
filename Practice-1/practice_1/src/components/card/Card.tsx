import React from "react";
import CardTitle from "../title/CardTitle";
import { CardPrice } from "./CardPrice";
import CardText from "./CardText";
import Button from "../button/Button";
import "./card.css"
import { CardProps } from "../../types/card";

export default function Card({
  className,
  title,
  unit,
  price,
  value1,
  value2,
  value3,
  value4,
  label,
}: CardProps): JSX.Element {
  return (
    <div className="card">
      <div className={`card__info card__info--${className}`}>
        <CardTitle className="name"><span>{title}</span></CardTitle>
        <CardPrice children1={unit} children2={price} ></CardPrice>
      </div>
      <div className="card__description">
        <CardText value1={value1} value2={value2} value3={value3} value4={value4}></CardText>
        <Button type="outline">{label}</Button>
      </div>
    </div>
  );
}