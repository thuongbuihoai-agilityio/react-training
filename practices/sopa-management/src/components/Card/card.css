.card {
  width: var(--w-medium);
  height: var(--h-large);
  border-radius: var(--rounded-primary);
  background: var(--bg-primary);
  margin-bottom: var(--m-large);
}

.card-image {
  text-align: center;
  margin-top: var(--m-5xl);
}

.card-item {
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  height: var(--h-medium);
  padding: var(--px-medium) var(--py-normal) var(--py-normal) var(--py-normal);
}

.card-description {
  display: flex;
  margin-top: var(--m-normal);
  justify-content: space-between;
}

.card-name {
  color: var(--dark);
  font-size: var(--text-extra-sm);
  line-height: var(--leading-primary);
  font-weight: var(--font-medium);
  text-decoration: none;
}

.card-image-item {
  object-fit: cover;
}
