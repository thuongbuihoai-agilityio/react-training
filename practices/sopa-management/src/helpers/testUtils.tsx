import { render } from '@testing-library/react';
import { createMemoryHistory } from 'history';
import {
  QueryClient,
  QueryClientProvider
} from 'react-query';
import {
  BrowserRouter,
  Router
} from 'react-router-dom';

export const renderRouterTest = (component: JSX.Element) => {
  const history = createMemoryHistory();
  return render(
    <Router location={history.location} navigator={history}>
      {component}
    </Router>
  );
};

export const renderWithRouterAndQuery = (children: JSX.Element) => {
  const queryClient = new QueryClient();

  return render(
    <BrowserRouter>
      <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
    </BrowserRouter>
  );
};
