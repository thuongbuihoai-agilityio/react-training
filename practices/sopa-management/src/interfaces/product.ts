export interface Product {
  id: string;
  name: string;
  price: number;
  color: string;
  image: {
    url: string;
    alt: string;
  };
  size: string;
  description?: string;
  quantity: number;
};
