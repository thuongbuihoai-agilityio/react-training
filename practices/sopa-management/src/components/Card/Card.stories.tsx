import type { Meta, StoryObj } from '@storybook/react';
import { BrowserRouter } from 'react-router-dom';

// Constants
import { IMAGE } from '@constants/image';

// Components
import Card from '@components/Card';

const meta: Meta<typeof Card> = {
  component: Card,
  decorators: [
    (Story) => (
      <BrowserRouter>
        <Story />
      </BrowserRouter>
    )
  ]
};

export default meta;
type Story = StoryObj<typeof Card>;

export const TextNormal: Story = {
  args: {
    name: 'Black shoes',
    color: 'Black',
    price: 145,
    href: '/',
    src: IMAGE.blackShoes,
    alt: IMAGE.altBlackShoes
  }
};
