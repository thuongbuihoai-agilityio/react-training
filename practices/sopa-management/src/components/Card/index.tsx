import { memo } from 'react';
import { Link } from 'react-router-dom';

// Components
import {
  Image,
  Price,
  PriceType,
  Text,
  ThemeType
} from '@components/common';

// Styles
import './card.css';

interface CardProps {
  key?: string;
  name?: string;
  color?: string;
  price?: number;
  href: string;
  src: string;
  alt?: string;
}

const Card = ({
  name = '',
  color = '',
  price = 0,
  href,
  src,
  alt = ''
}: CardProps) => (
  <div data-testid='card' className='card'>
    <div className='card-item'>
      <div className='card-image'>
        <Image
          href={href}
          url={src}
          alt={alt}
          width={322}
          height={130}
          className='card-image-item'
        />
      </div>
      <div className='card-info'>
        <Link to={href} className='card-name' aria-label={name}>
          {name}
        </Link>
        <div className='card-description'>
          <Text text={color} type={ThemeType.highlightPrimary} />
          <Price value={price} type={PriceType.primary} />
        </div>
      </div>
    </div>
  </div>
);

export default memo(Card);
