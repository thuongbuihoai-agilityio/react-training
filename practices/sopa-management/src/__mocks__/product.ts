import { Product } from "../interfaces/product";

const MOCK_PRODUCTS: Product[] = [
  {
    name: 'Black shoes',
    price: 87,
    size: 'Small',
    description:
      'Our goal in designing the super-soft Atoms Crew Socks was to provide a seamless match with your Atoms shoes, enhancing your comfort for everyday wear - making them the perfect socks for your everyday shoes.',
    color: 'Black',
    image: {
      url: '/images/products/black-shoes.png',
      alt: 'This is black shoes'
    },
    id: '1',
    quantity: 2
  },
  {
    name: 'Blue shoes',
    price: 45,
    size: 'Medium',
    description:
      'Our goal in designing the super-soft Atoms Crew Socks was to provide a seamless match with your Atoms shoes, enhancing your comfort for everyday wear - making them the perfect socks for your everyday shoes.',
    color: 'Blue',
    image: {
      url: '/images/products/blue-shoes.png',
      alt: 'This is blue shoes'
    },
    id: '2',
    quantity: 1
  }
];

const MOCK_PRODUCT: Product = {
  name: 'Black shoes',
  price: 87,
  size: 'Large',
  description: 'Our goal in designing the super-soft Atoms Crew Socks was to provide a seamless match with your Atoms shoes, enhancing your comfort for everyday wear - making them the perfect socks for your everyday shoes.',
  color: 'Black',
  image: {
    url: '/images/products/black-shoes.png',
    alt: 'This is black shoes'
  },
  id: '1',
  quantity: 3
};

export {
  MOCK_PRODUCTS,
  MOCK_PRODUCT
};
