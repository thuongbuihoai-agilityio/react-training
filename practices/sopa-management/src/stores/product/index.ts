import { create } from 'zustand';
import { Product } from '@interfaces/product';

interface ProductType {
  products: Product[];
  setProducts: (value: Product[]) => void;
}

export const useProductStore = create<ProductType>()((set) => ({
  products: [],
  setProducts: (data: Product[]) => set(() => ({ products: data }))
}));
