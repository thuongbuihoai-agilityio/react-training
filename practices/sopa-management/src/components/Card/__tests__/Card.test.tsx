// Libs
import '@testing-library/jest-dom';

// Mocks
import { MOCK_PRODUCT } from '@mocks/product';

// Components
import Card from '@components/Card';

// Helpers
import { renderRouterTest } from '@helpers/testUtils';

describe('Card component', () => {
  test('should render Card component', () => {
    const { getByTestId } = renderRouterTest(
      <Card
        href='/'
        src={MOCK_PRODUCT.image?.url}
        name={MOCK_PRODUCT.name}
        color={MOCK_PRODUCT.color}
        price={MOCK_PRODUCT.price}
      />
    );
    expect(getByTestId('card')).toBeInTheDocument();
  });

  test('should render Card component by default', () => {
    const { getByTestId } = renderRouterTest(
      <Card href='/' src={MOCK_PRODUCT.image?.url} />
    );
    expect(getByTestId('card')).toBeInTheDocument();
  });

  test('matches snapshot', () => {
    const { container } = renderRouterTest(
      <Card
        href='/'
        src={MOCK_PRODUCT.image?.url}
        name={MOCK_PRODUCT.name}
        color={MOCK_PRODUCT.color}
        price={MOCK_PRODUCT.price}
      />
    );
    expect(container).toMatchSnapshot();
  });
});
