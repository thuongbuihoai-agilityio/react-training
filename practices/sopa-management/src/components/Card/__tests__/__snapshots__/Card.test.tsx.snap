// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card component matches snapshot 1`] = `
<div>
  <div
    class="card"
    data-testid="card"
  >
    <div
      class="card-item"
    >
      <div
        class="card-image"
      >
        <a
          aria-label="link to home page"
          data-testid="image-link"
          href="/"
        >
          <img
            alt=""
            class="card-image-item"
            height="130"
            sizes=""
            src="/images/products/black-shoes.png"
            width="322"
          />
        </a>
      </div>
      <div
        class="card-info"
      >
        <a
          aria-label="Black shoes"
          class="card-name"
          href="/"
        >
          Black shoes
        </a>
        <div
          class="card-description"
        >
          <p
            class=" text-highlight-primary"
            data-testid="text"
          >
            Black
          </p>
          <p
            class=" price-primary"
            data-testid="price"
          >
            $
            87
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
